### PR TITLE
Rename Tag(s) to Event(s)

### PR DESCRIPTION
**Pull request by marbar3778**
- tag was replaced with event, but in some places it still mentions tag, would be easier to understand if we tried to replace it everywhere with event.

Signed-off-by: Marko Baricevic <marbar3778@yahoo.com>

* [ ] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -881,30 +881,30 @@ type TxIndexConfig struct {
...
+	// Comma-separated list of events to index (by default the only tag is "tx.hash")
 	//
 	// You can also index transactions by height by adding "tx.height" tag here.
 	//
-	// It's recommended to index only a subset of tags due to possible memory
+	// It's recommended to index only a subset of events due to possible memory
 	// bloat. This is, of course, depends on the indexer's DB and the volume of
 	// transactions.
-	IndexTags string `mapstructure:"index_tags"`
+	IndexEvents string `mapstructure:"index_events"`
```
Comment (melekes) [inline config/config.go]: problem here events is more complex data structure, so we'll need to explain how events can be transformed into `Comma-separated list of events` string

Comment (melekes) [reply config/config.go]: also, not clear what an event is in this context? just `Type` or `Type.TagName`

Comment (marbar3778) [reply config/config.go]: https://github.com/tendermint/tendermint/blob/master/abci/types/types.proto#L245 i believe tag should be replaced throughout the repo to event

**Comment by melekes** [comment]
https://github.com/tendermint/tendermint/pull/4077#discussion_r341856953

We need to agree on naming here. In #4077 I've proposed the new name - `compositeKey` (because it's composite of event.Type and event.AttributeName) for what we were previously calling`tag`. Let me know if you have a better idea!

**Comment by marbar3778** [comment]
I am fine with that, makes more sense, will make the adjustments,

File: config/config.go [hunk 2]
```diff
@@ -875,7 +875,7 @@ func (cfg *ConsensusConfig) ValidateBasic() error {
 // TxIndexConfig
 
 // TxIndexConfig defines the configuration for the transaction indexer,
-// including tags to index.
+// including compositeKeys to index.
```
Comment (melekes) [inline config/config.go]: I think we'll need to explain what compositeKey is here:

```
Remember that Event have a following structure:
type: [
  key: value,
  ...
]
```

composite key is a type and key, separated by a single dot. For example, given the Event below

```
"jack": [
  "account.number": 100
]
``` 

composite key will be `jack.account.number`

Comment (melekes) [inline config/config.go]:
```suggestion
// including composite keys to index.
```

File: docs/app-dev/indexing-transactions.md [hunk 3]
```diff
@@ -8,9 +8,9 @@ Tendermint allows you to index transactions and later query or subscribe
 to their results.
 
 Events can be used to index transactions and blocks according to what happened
-during their execution. Note that the set of events returned for a block from
+during their execution. Note that the set of keys returned for a block from
```
Comment (melekes) [inline docs/app-dev/indexing-transactions.md]:
```suggestion
during their execution. Note that the set of events returned for a block from
```

events is correct!

File: docs/app-dev/indexing-transactions.md [hunk 4]
```diff
@@ -8,9 +8,9 @@ Tendermint allows you to index transactions and later query or subscribe
...
 `BeginBlock` and `EndBlock` are merged. In case both methods return the same
-tag, only the value defined in `EndBlock` is used.
+key, only the value defined in `EndBlock` is used.
```
Comment (melekes) [inline docs/app-dev/indexing-transactions.md]:
```suggestion
type, only the key-value pairs defined in `EndBlock` are used.
```

File: docs/app-dev/indexing-transactions.md [hunk 5]
```diff
@@ -29,22 +29,22 @@ Let's take a look at the `[tx_index]` config section:
 #   2) "kv" (default) - the simplest possible indexer, backed by key-value storage (defaults to levelDB; see DBBackend).
 indexer = "kv"
 
-# Comma-separated list of tags to index (by default the only tag is "tx.hash")
+# Comma-separated list of compositeKeys to index (by default the only key is "tx.hash")
```
Comment (melekes) [inline docs/app-dev/indexing-transactions.md]:
```suggestion
# Comma-separated list of composite keys to index (by default the only key is "tx.hash")
```

File: docs/spec/abci/abci.md [hunk 6]
```diff
@@ -48,7 +48,7 @@ during the method's execution.
 Events can be used to index transactions and blocks according to what happened
 during their execution. Note that the set of events returned for a block from
 `BeginBlock` and `EndBlock` are merged. In case both methods return the same
-tag, only the value defined in `EndBlock` is used.
+event, only the value defined in `EndBlock` is used.
```
Comment (melekes) [inline docs/spec/abci/abci.md]:
```suggestion
type, only the key-value pairs defined in `EndBlock` are used.
```

File: docs/spec/abci/abci.md [hunk 7]
```diff
@@ -282,7 +282,7 @@ Commit are included in the header of the next block.
   - `ByzantineValidators ([]Evidence)`: List of evidence of
     validators that acted maliciously.
 - **Response**:
-  - `Tags ([]cmn.KVPair)`: Key-Value tags for filtering and indexing
+  - `Events ([]cmn.KVPair)`: Key-Value events for filtering and indexing
```
Comment (melekes) [inline docs/spec/abci/abci.md]: `[]cmn.KVPair` type is incorrect

```suggestion
  - `Events ([]Event)`: Events for filtering and indexing
```

File: docs/spec/abci/abci.md [hunk 8]
```diff
@@ -309,7 +309,7 @@ Commit are included in the header of the next block.
     be non-deterministic.
   - `GasWanted (int64)`: Amount of gas requested for transaction.
   - `GasUsed (int64)`: Amount of gas consumed by transaction.
-  - `Tags ([]cmn.KVPair)`: Key-Value tags for filtering and indexing
+  - `Events ([]cmn.KVPair)`: Key-Value events for filtering and indexing
```
Comment (melekes) [inline docs/spec/abci/abci.md]:
```suggestion
  - `Events ([]Event)`: Events for filtering and indexing
```

File: docs/spec/abci/abci.md [hunk 9]
```diff
@@ -337,7 +337,7 @@ Commit are included in the header of the next block.
...
```
Comment (melekes) [inline docs/spec/abci/abci.md]: same

File: docs/spec/abci/abci.md [hunk 10]
```diff
@@ -354,7 +354,7 @@ Commit are included in the header of the next block.
     voting power to 0 to remove).
   - `ConsensusParamUpdates (ConsensusParams)`: Changes to
     consensus-critical time, size, and other parameters.
-  - `Tags ([]cmn.KVPair)`: Key-Value tags for filtering and indexing
+  - `Events ([]cmn.KVPair)`: Key-Value events for filtering and indexing
```
Comment (melekes) [inline docs/spec/abci/abci.md]: same

File: libs/pubsub/query/query.go [hunk 11]
```diff
@@ -29,12 +29,12 @@ type Query struct {
 	parser *QueryParser
 }
 
-// Condition represents a single condition within a query and consists of tag
+// Condition represents a single condition within a query and consists of event
```
Comment (melekes) [inline libs/pubsub/query/query.go]:
```suggestion
// Condition represents a single condition within a query and consists of composite key
```

File: libs/pubsub/query/query.go [hunk 12]
```diff
@@ -63,7 +63,7 @@ func (q *Query) String() string {
 	return q.str
 }
 
-// Operator is an operator that defines some kind of relation between tag and
+// Operator is an operator that defines some kind of relation between event and
```
Comment (melekes) [inline libs/pubsub/query/query.go]:
```suggestion
// Operator is an operator that defines some kind of relation between composite key and
```

File: state/txindex/kv/kv.go [hunk 13]
```diff
@@ -27,31 +27,31 @@ var _ txindex.TxIndexer = (*TxIndex)(nil)
...
 	for _, o := range options {
 		o(txi)
 	}
 	return txi
 }
 
-// IndexTags is an option for setting which tags to index.
-func IndexTags(tags []string) func(*TxIndex) {
+// IndexEvents is an option for setting which events to index.
+func IndexEvents(events []string) func(*TxIndex) {
```
Comment (melekes) [inline state/txindex/kv/kv.go]:
```suggestion
func IndexEvents(compositeKeys []string) func(*TxIndex) {
```

File: state/txindex/kv/kv.go [hunk 14]
```diff
@@ -27,31 +27,31 @@ var _ txindex.TxIndexer = (*TxIndex)(nil)
...
+	txi := &TxIndex{store: store, compositeKeysToIndex: make([]string, 0), indexAllEvents: false}
 	for _, o := range options {
 		o(txi)
 	}
 	return txi
 }
 
-// IndexTags is an option for setting which tags to index.
-func IndexTags(tags []string) func(*TxIndex) {
+// IndexEvents is an option for setting which events to index.
```
Comment (melekes) [inline state/txindex/kv/kv.go]:
```suggestion
// IndexEvents is an option for setting which composite keys to index.
```

File: types/event_bus.go [hunk 15]
```diff
@@ -164,16 +164,16 @@ func (b *EventBus) PublishEventValidBlock(data EventDataRoundState) error {
 	return b.Publish(EventValidBlock, data)
 }
 
-// PublishEventTx publishes tx event with tags from Result. Note it will add
-// predefined tags (EventTypeKey, TxHashKey). Existing tags with the same names
+// PublishEventTx publishes tx event with events from Result. Note it will add
+// predefined keys (EventTypeKey, TxHashKey). Existing events with the same names
```
Comment (melekes) [inline types/event_bus.go]:
```suggestion
// predefined keys (EventTypeKey, TxHashKey). Existing events with the same keys
```

File: types/events.go [hunk 16]
```diff
@@ -126,7 +126,7 @@ type EventDataValidatorSetUpdates struct {
 ///////////////////////////////////////////////////////////////////////////////
 
 const (
-	// EventTypeKey is a reserved key, used to specify event type in tags.
+	// EventTypeKey is a reserved key, used to specify compositeKey type in events.
```
Comment (melekes) [inline types/events.go]:
```suggestion
	// EventTypeKey is a reserved composite key for event name.
```

File: types/event_bus_test.go [hunk 17]
```diff
@@ -29,7 +29,7 @@ func TestEventBusPublishEventTx(t *testing.T) {
 		},
 	}
 
-	// PublishEventTx adds all these 3 tags, so the query below should work
+	// PublishEventTx adds 3 compositeKeys, so the query below should work
```
Comment (melekes) [inline types/event_bus_test.go]:
```suggestion
	// PublishEventTx adds 3 composite keys, so the query below should work
```

File: config/toml.go [hunk 18]
```diff
@@ -347,22 +347,22 @@ peer_query_maj23_sleep_duration = "{{ .Consensus.PeerQueryMaj23SleepDuration }}"
 #   2) "kv" (default) - the simplest possible indexer, backed by key-value storage (defaults to levelDB; see DBBackend).
 indexer = "{{ .TxIndex.Indexer }}"
 
-# Comma-separated list of tags to index (by default the only tag is "tx.hash")
```
Comment (melekes) [inline config/toml.go]: what about "Remember that Event has the following structure"?

File: docs/app-dev/indexing-transactions.md [hunk 19]
```diff
@@ -10,12 +10,23 @@ to their results.
...
+An Event has a composite key associated with it. A `compositeKey` is constructed by its type and key separated by a dot.
+For example:
+
+```json
+"jack": [
+  "account.number": 100
+]
+```
+
+would be equal to the composite key of `jack.account.numer`.
```
Comment (melekes) [inline docs/app-dev/indexing-transactions.md]:
```suggestion
would be equal to the composite key of `jack.account.number`.
```

File: docs/tendermint-core/configuration.md [hunk 20]
```diff
@@ -299,22 +298,22 @@ blocktime_iota = "1s"
 #   2) "kv" (default) - the simplest possible indexer, backed by key-value storage (defaults to levelDB; see DBBackend).
 indexer = "kv"
 
-# Comma-separated list of tags to index (by default the only tag is "tx.hash")
```
Comment (melekes) [inline docs/tendermint-core/configuration.md]: let's put "Remember that Event has the following structure" here too

File: state/txindex/kv/kv_bench_test.go [hunk 21]
```diff
@@ -25,7 +25,7 @@ func BenchmarkTxSearch(b *testing.B) {
 	}
 
 	allowedTags := []string{"transfer.address", "transfer.amount"}
-	indexer := NewTxIndex(db, IndexTags(allowedTags))
+	indexer := NewTxIndex(db, IndexEvents(allowedTags))
```
Comment (melekes) [inline state/txindex/kv/kv_bench_test.go]:
```suggestion
	indexer := NewTxIndex(db, IndexEvents(allowedKeys))
```

File: state/txindex/kv/kv_bench_test.go [hunk 22]
```diff
@@ -25,7 +25,7 @@ func BenchmarkTxSearch(b *testing.B) {
 	}
 
 	allowedTags := []string{"transfer.address", "transfer.amount"}
```
Comment (melekes) [inline state/txindex/kv/kv_bench_test.go]:
```suggestion
	allowedKeys := []string{"transfer.address", "transfer.amount"}
```

File: config/toml.go [hunk 23]
```diff
@@ -348,6 +348,7 @@ peer_query_maj23_sleep_duration = "{{ .Consensus.PeerQueryMaj23SleepDuration }}"
 indexer = "{{ .TxIndex.Indexer }}"
 
 # Comma-separated list of compositeKeys to index (by default the only key is "tx.hash")
+# Remember that Event has the following structure: key.type
 #
```
Comment (melekes) [inline config/toml.go]: let's copy the actual structure of the Event type too

File: docs/tendermint-core/configuration.md [hunk 24]
```diff
@@ -299,6 +299,7 @@ blocktime_iota = "1s"
 indexer = "kv"
 
 # Comma-separated list of compositeKeys to index (by default the only key is "tx.hash")
+# Remember that Event has the following structure: key.type
 #
```
Comment (melekes) [inline docs/tendermint-core/configuration.md]: here also

**Comment by melekes** [comment]
we will need a changelog entry since we're breaking a config

File: CHANGELOG_PENDING.md [hunk 25]
```diff
@@ -70,14 +72,14 @@ program](https://hackerone.com/tendermint).
 - Blockchain Protocol
 
   - [abci] \#2521 Remove `TotalTxs` and `NumTxs` from `Header`
-  - [types] [\#4151](https://github.com/tendermint/tendermint/pull/4151) Enforce ordering of votes in DuplicateVoteEvidence to be lexicographically sorted on BlockID
+  - [types][\#4151](https://github.com/tendermint/tendermint/pull/4151) Enforce ordering of votes in DuplicateVoteEvidence to be lexicographically sorted on BlockID
```
Comment (melekes) [inline CHANGELOG_PENDING.md]: 🤣 so I am adding white spaces and you're deleting them, nice

Comment (melekes) [inline CHANGELOG_PENDING.md]: 🤣 so I am adding white spaces and you're deleting them, nice

Comment (marbar3778) [reply CHANGELOG_PENDING.md]: ah sorry this is my editor, will turn it off

**Pull request merged**